src/util/db/dbid.h

```diff
@@ -1,126 +1,138 @@
 #pragma once
 
-
-#include <functional>
-#include <ostream>
-#include <utility>
-
 #include <QString>
 #include <QVariant>
 #include <QtDebug>
+#include <functional>
+#include <ostream>
+#include <utility>
 
 #include "util/assert.h"
 
-
-// Base class for ID values of objects that are stored in the database.
+/// Base class for ID values of objects that are stored in the database.
 //
-// Implicit conversion from/to the native value type has been disabled
-// on purpose in favor of explicit conversion. The internal value type
-// is also hidden as an implementation detail.
+/// Implicit conversion from/to the native value type has been disabled
+/// on purpose in favor of explicit conversion. The internal value type
+/// is also hidden as an implementation detail.
 //
-// Although used as a base class this class has a non-virtual destructor,
-// because destruction is trivial and for maximum efficiency! Derived
-// classes may add additional behavior (= member functions), but should
-// not add any additional state (= member variables). Inheritance is
-// only needed for type-safety.
+/// Although used as a base class this class has a non-virtual destructor,
+/// because destruction is trivial and for maximum efficiency! Derived
+/// classes may add additional behavior (= member functions), but should
+/// not add any additional state (= member variables). Inheritance is
+/// only needed for type-safety.
 class DbId {
-protected:
-public:
-    // Alias for the corresponding native type. It keeps the
-    // implementation of this class flexible if we ever gonna
-    // need to change it from 'int' to 'long' or any other type.
+  public:
+    /// Alias for the corresponding native type. It keeps the
+    /// implementation of this class flexible if we ever gonna
+    /// need to change it from 'int' to 'long' or any other type.
     typedef int value_type;
 
-    DbId()
-        : m_value(kInvalidValue) {
-        DEBUG_ASSERT(!isValid());
+    static constexpr value_type kInvalidValue = -1;
+    static constexpr value_type kMinValue = 0;
+
+    static constexpr bool isValidValue(value_type value) {
+        static_assert(kInvalidValue < kMinValue);
+        return value >= kMinValue;
+    }
+
+    /// Type tag for overloaded constructor that doesn't assert
+    struct NoAssert {};
```

Comment:
That looks like an ugly hack to me, that is at least hard to understand for no-c++-experts.  

An empty struct has a size of one byte, because every C++ object has an address.  So you may make it explicit and pass a real byte or bool and check it in the assertion as well. I am sure the compiler will optimize it out even in the debug version. 

The alternative would be to inherit a non asserting version.

Reply:
Please sketch your solution if you know a better one.

This one is simple and is resolved at compile time. Empty structs are commonly used in template meta-programming for type dispatching. This is what overloading does.

Reply:
Something like this: 

```
    constexpr DbId()
            : m_value(kInvalidValue) {
    }
    constexpr DbId(value_type value, bool assertDefinedValues = true)
            : m_value(value) {
        Q_UNUSED(assertDefinedValues)    
        DEBUG_ASSERT(!assertDefinedValues || isValid() || m_value == kInvalidValue);
    }
    explicit DbId(QVariant variant)
            : DbId(valueOf(std::move(variant)), false) {
    }
```

Reply:
On a second thought I don't understand why we need the path without the assertion at all. I think there is no use case for allowing -5 or such.
Us there one?

Reply:
constexpr + testing

Reply:
And no, I won't add such an ugly and error prone runtime hack with a boolean. It completely misses the point.

Reply:
> constexpr + testing

Please explain, IMHO nothing of this rectifies using an empty struct.

I would prefer to remove the path without the assertion and keep the code simple and maintainable.

Reply:
We need the path for testing.